app/src/main/java/com/x8bit/bitwarden/data/autofill/fido2/manager/Fido2CredentialManagerImpl.kt

```diff
@@ -136,40 +134,52 @@ class Fido2CredentialManagerImpl(
         val clientData = request.clientDataHash
             ?.let { ClientData.DefaultWithCustomHash(hash = it) }
             ?: ClientData.DefaultWithExtraData(androidPackageName = callingAppInfo.getAppOrigin())
-        val origin = request.origin
+        val origin = callingAppInfo.origin
             ?: getOriginUrlFromAssertionOptionsOrNull(request.requestJson)
             ?: return Fido2CredentialAssertionResult.Error
+        val relyingPartyId = json
+            .decodeFromStringOrNull<PasskeyAssertionOptions>(request.requestJson)
+            ?.relyingPartyId
+            ?: return Fido2CredentialAssertionResult.Error
 
-        return vaultSdkSource
-            .authenticateFido2Credential(
-                request = AuthenticateFido2CredentialRequest(
-                    userId = userId,
-                    origin = origin,
-                    requestJson = """{"publicKey": ${request.requestJson}}""",
-                    clientData = clientData,
-                    selectedCipherView = selectedCipherView,
-                    isUserVerificationSupported = true,
-                ),
-                fido2CredentialStore = this,
-            )
-            .map { it.toAndroidFido2PublicKeyCredential() }
-            .mapCatching { json.encodeToString(it) }
-            .fold(
-                onSuccess = { Fido2CredentialAssertionResult.Success(it) },
-                onFailure = { Fido2CredentialAssertionResult.Error },
-            )
+        val validateOriginResult = validateOrigin(
+            callingAppInfo = callingAppInfo,
+            relyingPartyId = relyingPartyId,
+        )
+
+        return when (validateOriginResult) {
```

Comment:
💭 Non-blocking, but looking at this and the similar logic in the VaultItemListingVM wonderings if some of this logic could be spit out to another function and the VM could just call it?

Reply:
Good call. I have changes on deck that extract `GetCredentials` and `AuthenticateCredentials` requests into a dedicated View that will clean a lot of this up. I'll keep this in mind with those changes. 👍